src/main/java/net/minecraftforge/network/ConnectionData.java

```diff
@@ -66,4 +90,76 @@
     {
         return channels;
     }
+
+    /**
+     * A class for holding the mod mismatch data of a failed handshake.
+     * Contains a list of mismatched channels, the channels present on the side the handshake failed on, the mods with mismatching registries (if available) and the information of whether the mismatching data's origin is the server.
+     * @param mismatchedChannelData The ids and respective mod versions of the mismatched or missing channels
+     * @param presentChannelData The ids and respective mod names and versions of the present channels corresponding to the mismatched channels, stored like this: [channelId -> [modName, modVersion]]
+     * @param missingRegistryModData The ids and names of mods that have mismatched registry data
+     */
+    public record ModMismatchData(Map<ResourceLocation, String> mismatchedChannelData, Map<ResourceLocation, Pair<String, String>> presentChannelData, Map<String, String> missingRegistryModData, boolean mismatchedDataFromServer)
```

Comment:
In this class you should use `Map.of` for empty maps instead of new `HashMap`s every time.